File: Algorithm-Repository/Basic-Arithmetic/Cpp/GCD.cpp
// GCD.cpp
// Language: C++
// Purpose: Compute the Greatest Common Divisor (GCD) of two integers using Euclidean Algorithm
// Time Complexity: O(log(min(a, b)))
// Space Complexity: Iterative: O(1), Recursive: O(log(min(a, b))) due to recursion stack

/*
 * ✅ Description:
 * GCD (Greatest Common Divisor) finds the largest integer that divides both numbers without remainder.
 * Includes both iterative and recursive implementations for clarity.
 *
 * 🧼 Clean and optimized implementation:
 *   - Handles negative numbers using llabs()
 *   - Works with large integers using long long
 *
 * 🧪 Sample Input / Output:
 *   Input:
 *     48 18
 *   Output:
 *     GCD of 48 and 18 is: 6
 *
 * 🧰 Use Case:
 *   - Simplifying fractions
 *   - Computing LCM (lcm(a,b) = |a*b| / gcd(a,b))
 *   - Modular arithmetic, cryptography (modular inverse, RSA)
 */


#include <bits/stdc++.h>
using namespace std;

long long gcd(long long a, long long b){
     a = llabs(a);
    b = llabs(b);

    while(b !=0){
        long long temp =b;
        b = a%b;
        a = temp;

    }
    return a;

}

// recursion approach
long long gcdRecursive(long long a, long long b) {
    if (b == 0) return llabs(a);
    return gcdRecursive(b, a % b);
}
int main(){
    long long a, b;
    cin >> a >> b;

     cout << "GCD(" << a << ", " << b << ") = " << gcd(a, b) << "\n";
    return 0;
}

File: Algorithm-Repository/Basic-Arithmetic/Java/GCD.java
// LCM.java
// Language: Java
// Purpose: Compute the Least Common Multiple (LCM) of two integers using GCD
// Time Complexity: O(log(min(a, b))) — dominated by GCD computation
// Space Complexity: O(1)
/*
 * ✅ Description:
 * LCM (Least Common Multiple) finds the smallest positive integer that is divisible by both numbers.
 * Uses the formula: lcm(a, b) = |a * b| / gcd(a, b)
 *
 * 🧼 Clean and optimized implementation:
 *   - Handles negative numbers
 *   - Uses gcd function to ensure efficiency
 *   - Works for large integers using long
 *
 * 🧪 Sample Input / Output:
 *   Input:
 *     12 18
 *   Output:
 *     LCM of 12 and 18 is: 36
 *
 * 🧰 Use Case:
 *   - Scheduling problems (finding common intervals)
 *   - Simplifying fractions
 *   - Problems involving multiples and modular arithmetic
 */
import java.util.Scanner;
public class GCD {
    static long gcd(long a, long b) {
        while (b != 0) {
            long temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }
    static long gcdRecursive(long a, long b) {
        if (b == 0)
            return a;
        return gcd(b, a % b);
    }
    public static void main(String[] args) {
        long a,b;
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter two numbers: ");
        a = sc.nextLong();
        b = sc.nextLong();
        System.out.println("GCD of "+a+" and "+b+" is: "+gcd(a,b));
        sc.close();
        
    }
}
